vm/src/stdlib/array.rs

```diff
@@ -223,19 +225,127 @@ macro_rules! def_array_enum {
                 }
             }
 
-            fn setitem(&mut self, needle: Either<isize, PySliceRef>, value: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
-                match needle {
-                    Either::A(i) => {
-                        let i = self.idx(i, "array assignment", vm)?;
-                        match self {
-                            $(ArrayContentType::$n(v) => { v[i] = TryFromObject::try_from_object(vm, value)? },)*
+            fn setitem_by_slice(&mut self, slice: PySliceRef, value: ArrayContentType, vm: &VirtualMachine) -> PyResult<()> {
+                let step = slice.step_index(vm)?.unwrap_or_else(BigInt::one);
+                let start = slice.start_index(vm)?;
+                let stop = slice.stop_index(vm)?;
+                if step.is_zero() {
+                    return Err(vm.new_value_error("slice step cannot be zero".to_owned()));
+                }
+                match self {
+                    $(ArrayContentType::$n(elements) => {
+                        if let ArrayContentType::$n(items) = value {
+                            if step == BigInt::one() {
+                                let mut range = get_slice_range(&start, &stop, elements.len());
+                                if range.end < range.start {
+                                    range.end = range.start;
+                                }
+                                elements.splice(range, items);
+                                return Ok(());
+                            }
+                            let (start, stop, step, is_negative) = if step.is_negative() {
+                                (
+                                    stop.map(|x| {
+                                        if x == (-1).to_bigint().unwrap() {
```

Comment:
instead of making a new bigint, how about
```rust
if x.to_i32() == Some(-1) {
```
To use bigint, `one()` is still available here too
```rust
if x == -BigInt::one() {
```